test/common/http/http2/hpack_fuzz_test.cc

```diff
@@ -0,0 +1,163 @@
+}
+
+std::vector<nghttp2_nv> decodeHeaders(nghttp2_hd_inflater* inflater,
+                                      const Buffer::OwnedImpl& payload, bool end_headers) {
+  // Decode using nghttp2
+  Buffer::RawSliceVector slices = payload.getRawSlices();
+  const int num_slices = slices.size();
+  ASSERT(num_slices == 1, absl::StrCat("number of slices ", num_slices));
+
+  std::vector<nghttp2_nv> decoded_headers;
+  int inflate_flags = 0;
+  nghttp2_nv decoded_nv;
+  while (slices[0].len_ > 0) {
+    ssize_t result = nghttp2_hd_inflate_hd2(inflater, &decoded_nv, &inflate_flags,
+                                            reinterpret_cast<uint8_t*>(slices[0].mem_),
+                                            slices[0].len_, end_headers);
+    // Decoding should not fail and data should not be left in slice.
+    ASSERT(result >= 0);
+
+    slices[0].mem_ = reinterpret_cast<void*>(reinterpret_cast<uint8_t*>(slices[0].mem_) + result);
+    slices[0].len_ -= result;
+
+    if (inflate_flags & NGHTTP2_HD_INFLATE_EMIT) {
+      // One header key value pair has been successfully decoded.
+      decoded_headers.push_back(decoded_nv);
+    }
+  }
+
+  if (end_headers) {
+    nghttp2_hd_inflate_end_headers(inflater);
+  }
+
+  return decoded_headers;
+}
+
+int nvCompare(const void* a_in, const void* b_in) {
+  const nghttp2_nv* a = reinterpret_cast<const nghttp2_nv*>(a_in);
+  const nghttp2_nv* b = reinterpret_cast<const nghttp2_nv*>(b_in);
+
+  absl::string_view a_str(reinterpret_cast<char*>(a->name), a->namelen);
+  absl::string_view b_str(reinterpret_cast<char*>(b->name), b->namelen);
+  if (a_str > b_str) {
+    return 1;
+  }
+  if (a_str < b_str) {
+    return -1;
+  }
+  return 0;
+}
+
+DEFINE_PROTO_FUZZER(const test::common::http::http2::HpackTestCase& input) {
+  // Validate headers.
+  try {
+    TestUtility::validate(input);
+  } catch (const EnvoyException& e) {
+    ENVOY_LOG_MISC(trace, "EnvoyException: {}", e.what());
+    return;
+  }
+
+  // Create name value pairs from headers.
+  std::vector<nghttp2_nv> input_nv = createNameValueArray(input.headers());
+  // Skip encoding empty headers. nghttp2 will throw a nullptr error on runtime if it receives a
+  // nullptr input.
+  if (!input_nv.data()) {
+    return;
+  }
+
+  // Create Deflater and Inflater
+  nghttp2_hd_deflater* deflater = nullptr;
+  int rc = nghttp2_hd_deflate_new(&deflater, kHeaderTableSize);
+  ASSERT(rc == 0);
+  nghttp2_hd_inflater* inflater = nullptr;
+  rc = nghttp2_hd_inflate_new(&inflater);
+  ASSERT(rc == 0);
+
+  // Encode headers with nghttp2.
+  const Buffer::OwnedImpl payload = encodeHeaders(deflater, input_nv);
+  ASSERT(!payload.getRawSlices().empty());
+
+  // Decode headers with nghttp2
+  std::vector<nghttp2_nv> output_nv = decodeHeaders(inflater, payload, input.end_headers());
+
+  // Verify that decoded == encoded.
+  ASSERT(input_nv.size() == output_nv.size());
+  std::qsort(input_nv.data(), input_nv.size(), sizeof(nghttp2_nv), nvCompare);
```

Comment:
Nit: it's probably preferred to use C++ sort rather than the the C style comparator and qsort function with sizeof etc.
```
std::sort(input_nv.begin(), input_nv.end(), NvComparator());
```
NvComparator would be slightly different than the -1/0/1 return from your C comparator; its `operator()` would just return the 'less-than' result.

Reply:
nice, that cleaned up things a bit!